Feature/experimental design

lukaszimmermann opened the pull request:
Contains the most recent version of the experimental design, as implemented on the MSstats and ProteomicsLFQ branch. Most implementation was moved from `ExperimentalDesign.h` file to the `ExperimentalDesign.cpp` file.

Also include updated information about the sample section in `ExperimentalDesign.h`.

hroest commented:
I miss the unit tests and the pyOpenMS bindings. You fix the bindings but you dont provide new ones for the new classes 

@timosachsenberg @jpfeuffer should we have strict rules for this?
The rules for having tests for each class and public method are pretty clear but the rules for Python bindings are not. However, if it is part of the public API it should have Python bindings IMHO

hroest commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
 
-    for (TextFile::ConstIterator sit = tf.begin(); sit != tf.end(); ++sit)
+    // Parse Error of filename if test holds
+    void parseErrorIf(const bool test, const String &filename, const String &message)
```
again, this should be in `FORMAT`

hroest commented:
To me it seems like the class does too much and should be split up into a File IO class in `FORMAT` and a logical class that manages the data. Also, we need unit tests for these classes

timosachsenberg commented on src/openms/include/OpenMS/METADATA/ExperimentalDesign.h:
```diff
@@ -53,51 +61,270 @@ namespace OpenMS
+      SampleSection() = default;
+
+      // Number of columns of the Sample Section
+      Size n_columns_;
```
do we need n_columns_? if so it should not be public

timosachsenberg commented on src/openms/include/OpenMS/METADATA/ExperimentalDesign.h:
```diff
@@ -53,51 +61,270 @@ namespace OpenMS
+
+      // The entries of the Sample Section, filled while parsing
+      // the Experimental Design File
+      std::vector< std::vector < String > > content_;
```
should be private? same for  other members

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
-      // run-level rows
-      StringList cells;
-      s.split("\t", cells);
+    inline bool contains(const StringList &vec, const String &item)
```
otherwise this could be part of the list classes (was it ListUtils?)

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
-        throw Exception::ParseError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, tsv_file,
-          "Error: Not all columns of experimental design provided: " + String(s) + ".");
-      }
+    void parse_header(const StringList &header,
```
parse_header should be parseHeader_ and private

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
       }
 
-      if (!File::exists(r.file))
+      /* Currently, existence of file is not required
```
check if commented section is needed

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
+
+      // Maps the column header string to the column index for
+      // the run section
+      std::map <String, Size> _run_column_header_to_index;
```
we don't prefix _ to variable names

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
+      Size n_col = 0;
+
+      const TextFile text_file(tsv_file, true);
+      for (String s : text_file) {
```
{ into new line

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
+        trimAll(cells);
+
+        // State = 0 : Run Section header
+        if (state == 0) {
```
consider using enums for states

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
-    if (frac2run.size() > 1)
+    map<pair<String, unsigned>, unsigned> ExperimentalDesign::pathChannelMapper(
+            const bool basename,
+            unsigned (*f)(const ExperimentalDesign::RunRow &row)) const
```
why does it need to be that generic?

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
+
+    /* Implementations of SampleSection */
+
+    void ExperimentalDesign::SampleSection::getSamples(std::set<unsigned> &samples) const
```
better to just return the set of samples?
filling the referenced vector is kind of old-style c++

timosachsenberg commented on src/openms/source/METADATA/ExperimentalDesign.cpp:
```diff
@@ -41,152 +41,578 @@
+      }
+    }
+
+    void ExperimentalDesign::SampleSection::getFactors(std::set< String > &factors) const
```
same here...

timosachsenberg commented:
@lukaszimmermann looks pretty good. I agree with @hroest that we should separate the parsing and storing from the rest (and move that class to FORMAT).
Also please check the accessibility of the members (I think many of them can be private).
Regarding how mandatory pyOpenMS bindings are: I think this is a good topic for the developer retreat.

lukaszimmermann commented:
Thanks a lot for the feedback. I will split the class into IO and structure parts.

lukaszimmermann commented:
I addressed most of the points. It is not entirely clear to me how the class `ExperimentalDesignTable` can be split. The class does three things:
- Load Experimental design from various formats (static factory methods)
- Sanity checking of the format of the experimental design
- Providing several conventient accessors.

timosachsenberg commented:
For consistency reasons, it would make sense to move the IO only part (only the tsv reading) to an IO class .
Regarding the name ExperimentalDesignTable that @hroest proposed - I am not 100% happy with that.
For one, these are actually two tables in the file. For another, the tabular representation is the representation in the file. Otherwise one could argue that we have several classes named without the suffix *Table. The later is not that important to me but naming it ExperimentalDesign consisting of a RunSection and a SampleSection seemed not a worse choice in the first place. If one wants to emphasize the tabular nature one could rename RunSection to RunTable and SampleSection to SampleTable.
Again, not that important to me but maybe it was not clear that there are (up to) two tables in the design.

timosachsenberg commented:
@lukaszimmermann moving the IO to a separate class makes it easier to, e.g. modularize the IO part in OpenMS. We don't to this right now but in the future, the IO stuff could be a namespace thing, a separate link library, or a c20 and beyond module.

lukaszimmermann commented:
I moved the factory methods to the `ExperimentalDesignIO` class (which now has the `load` and `from*` static members).

timosachsenberg commented:
Hi Lukas, that was quick.
Two things:
1. We usually name the IO classes with suffix File
2. static ExperimentalDesign fromConsensusMap(const ConsensusMap& c); I would not put general conversion functions in a IO class

timosachsenberg commented:
also consider if friendship is needed

lukaszimmermann commented:
@timosachsenberg 
Ok, renamed `ExperimentalDesignIO` to `ExperimentalDesignFile` and put the converter methods back to `ExperimentalDesign`. 

Removed friend declarations.

Shall were remove the default constructors and setters? Then we could make the `ExperimentalDesign` class immutable.

timosachsenberg commented:
Good idea but I think we need the default constructor for pyOpenMS

hroest commented on src/openms/include/OpenMS/FORMAT/ExperimentalDesignFile.h:
```diff
@@ -0,0 +1,69 @@
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_KERNEL_EXPERIMENTALDESIGNFILE_H
+#define OPENMS_KERNEL_EXPERIMENTALDESIGNFILE_H
```
change KERNEL to FORMAT ?

hroest commented on src/openms/include/OpenMS/FORMAT/ExperimentalDesign.h:
```diff
@@ -0,0 +1,271 @@
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_KERNEL_EXPERIMENTALDESIGN_H
+#define OPENMS_KERNEL_EXPERIMENTALDESIGN_H
```
this could go either into METADATA, but probably not into FORMAT

lukaszimmermann commented:
ExperimentalDesign.h was moved to METADATA

hroest commented:
@lukaszimmermann your pyOpenMS compilation failed, see https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms_pr_docker_pyopenms/75/console

hroest commented:
@lukaszimmermann please fix the following error:

```
git clone https://github.com/hroest/docker_files.git && \
cd docker_files/openms/pyopenms && \
docker build --build-arg branch=feature/experimental_design --build-arg repo=lukaszimmermann/OpenMS .
[...]
Level 25:root:parsing progress 200 out of 483
Level 25:root:parsing progress 250 out of 483
Level 25:root:parsing progress 300 out of 483
Level 25:root:parsing progress 350 out of 483

Error compiling Cython file:
------------------------------------------------------------
...

cdef extern from "<OpenMS/METADATA/ExperimentalDesign.h>" namespace "OpenMS":
    
    cdef cppclass ExperimentalDesign "OpenMS::ExperimentalDesign":
        ExperimentalDesign(ExperimentalDesign) nogil except + #wrap-ignore
        libcpp_vector[ ExperimentalDesign_RunRow] rows
                                               ^
------------------------------------------------------------

/OpenMS/src/pyOpenMS/pxds/ExperimentalDesign.pxd:8:48: unknown type in template argument

```

hroest commented on src/openms/include/OpenMS/METADATA/ExperimentalDesign.h:
```diff
@@ -34,70 +34,179 @@
+    void setRunSection(const RunRows& run_section);
+
+    // Returns the Sample Section of the experimental design file
+    const ExperimentalDesign::SampleSection& getSampleSection() const;
```
I think the method names and documentation should be more clear

timosachsenberg commented on src/openms/include/OpenMS/METADATA/ExperimentalDesign.h:
```diff
@@ -34,70 +34,179 @@
-    /// TODO: add possibility to provide optional columns that map to additional run-level specific meta data (see, for example, the mzTab specification)
-    class OPENMS_DLLAPI MSRun
+
+    class OPENMS_DLLAPI RunRow
```
For multiplexed data one row does not necessarily correspond to an injection.
The name run section and sample section have been used to distinguish the
meta data associated with mapping quant. channels in fractionated files
and the sample specific meta data (like conditions etc.)

timosachsenberg replied:
we could also call this MSFileSection and MSFileRow? Would this be better?

hroest replied:
ok, in that case I think it should just be documented better. I dont really like the naming of "rows" and "columns" here since it referers to a specific file layout instead of concepts used by experimentalisits

hroest commented:
@lukaszimmermann please merge https://github.com/lukaszimmermann/OpenMS/pull/1 for pyOpenMS to compile

timosachsenberg commented on src/openms/include/OpenMS/FORMAT/ExperimentalDesignFile.h:
```diff
@@ -0,0 +1,125 @@
+  ///    Required to process fractionated data. One Column must be 'Sample', other columns
+  ///    are unspecified and can contain arbitrary factors
+
+ Sample	Some_Condition	Technical_Replicate
```
what about Technical_Replicate? is this required by MSStats? if so this should be documented

timosachsenberg replied:
Also it should be documented if something is an id used for grouping (e.g., all technical replicates of a sample have the same number) or for enumeration (e.g., 1,2,3 to denote TR1, TR2, and TR3)

lukaszimmermann replied:
The columns in the Sample table are arbitrary, you can name them whatever you like (with the exception of the Sample column, of course). But you need to tell MSstatsConverter, which column in the Sample table holds the Biological Replicate. This is a parameter of the converter. There is a default value for this parameter.

timosachsenberg replied:
ok so it should be documented if Bio.Rep. is a grouping identifier or a counting identifier for MSstats

timosachsenberg commented on src/openms/include/OpenMS/FORMAT/ExperimentalDesignFile.h:
```diff
@@ -0,0 +1,125 @@
+                                      SILAC with light and heavy: 1..2
+         Sample:                      Index of sample measured in the specified channel X, in fraction Y of run Z
+
+	Fraction Group	Fraction	Path(Spectra File)	Channel		Sample
```
done

timosachsenberg commented:
TODO: clean up code required. move implementations to cpp file

hroest commented on src/tests/class_tests/openms/source/ExperimentalDesign_test.cpp:
```diff
@@ -75,115 +76,118 @@ END_SECTION
 START_SECTION((bool sameNrOfMSFilesPerFraction() const ))
 {
-  // TODO
+  bool b = design.sameNrOfMSFilesPerFraction(); 
```
please also test whether it returns the right value!!

hroest commented on src/tests/class_tests/openms/data/ExperimentalDesign_input_1.tsv:
```diff
@@ -0,0 +1,29 @@
+Fraction_Group	Fraction	Spectra_Filepath	Label	Sample
+1	1	JD_06232014_sample1-A.raw	1	1
+2	1	JD_06232014_sample1_B.raw	1	2
+3	1	JD_06232014_sample1_C.raw	1	3
```
@timosachsenberg can you please make an example file that covers multiple cases, e.g. multiple fractionations or a fractionation combined with TMT ? so that people have some idea what such a file would look like

timosachsenberg replied:
done in ExperimentalDesign_input_2.tsv fractionated 4plex

hroest commented on src/tests/class_tests/openms/source/ExperimentalDesignFile_test.cpp:
```diff
@@ -0,0 +1,79 @@
+START_SECTION((static ExperimentalDesign load(const String &tsv_file, bool require_spectra_files)))
+{
+ExperimentalDesign design = ExperimentalDesignFile::load(
+  OPENMS_GET_TEST_DATA_PATH("ExperimentalDesign_input_1.tsv"), false);
```
please test here that the data was loaded properly, e.g. all the fields are populated with the right values

timosachsenberg replied:
done in ExperimentalDesign_test where I load the same files but test the content

hroest commented:
starts to look good, however your unit tests should not just test that you can call the function but also test taht the function does what it is supposed to do. Think of it as a way of documenting your code ...

timosachsenberg commented:
sure ... still stubs. I push frequently to keep Lukas posted
